File: client/client_fetch_id.py
#!/usr/bin/env python3

import sys
import httplib2
import time
import json
import urllib

SERVER_URL = 'http://gaisq.cs.ccu.edu.tw:4567'

class YoutubeRelatedFetcher:
	def __init__(self):
		self.conn = httplib2.Http()

	def get_related_id(self, vid):
		while True:
			try:
				res, content = self.conn.request("http://gdata.youtube.com/feeds/api/videos/" + vid + "/related?max-results=50&fields=entry(id)&alt=json")
				content = content.decode('utf-8')
				break
			except:
				self.conn = httplib2.Http()
		
		if res.status != 200:
			if content.find("too_many_recent_calls") != -1:
				#ban by server or no video
				raise IOError
			else:
				return False
		else:
			data = json.loads(content)
			related_id = []
			for t in data['feed']['entry']:
				related_id.append(t['id']['$t'].split('/')[-1])
			return related_id

def get_vids():
	conn = httplib2.Http()
	resp, content = conn.request("{}/youtube/related_ids/?limit={}".format(SERVER_URL, 100))
	if resp.status == 200:
		return json.loads(content.decode("utf-8"))
	else:
		return []

def post_vids(vids):
	print("post vids, size = {}".format(len(vids)))
	conn = httplib2.Http()
	data = {'ids': json.dumps(vids)}
	resp, content = conn.request("{}/youtube/ids/".format(SERVER_URL), "POST", urllib.parse.urlencode(data))

def post_vids_fetched(vids):
	print("post vids fetched, size = {}".format(len(vids)))
	conn = httplib2.Http()
	data = {'ids': json.dumps(vids)}
	resp, content = conn.request("{}/youtube/related_ids/".format(SERVER_URL), "POST", urllib.parse.urlencode(data))

if __name__ == '__main__':
	fetcher = YoutubeRelatedFetcher()
	while True:
		vids = get_vids()
		if len(vids) == 0:
			break
		related_ids = []
		for vid in vids:
			try:
				tmp = fetcher.get_related_id(vid)
			except:
				print("sleep")
				time.sleep(300)
			if tmp:
				related_ids.extend(tmp)

		post_vids(related_ids)
		post_vids_fetched(vids)

File: client/client_fetch_id.rb
#!/usr/bin/env ruby
# encoding: UTF-8

require 'typhoeus'
require 'json'
require 'logger'

log = Logger.new(STDOUT)
log.level = Logger::DEBUG
id_db = File.open('id_db', "a+")
id_need_fetch = File.open('id_need_fetch', "r")
id_fetched = File.open('id_fetched', "a+")

def parse_id(url)
  return url[42..52]
end

def parse_related_id(data)
  related_ids = []
  json = JSON.parse(data)

  return [] unless json['feed']['entry']
  json['feed']['entry'].each do |t|
    related_ids << parse_id(t['id']['$t'])
  end

  return related_ids
end

log.info("client start")
while true
  #read unfetched id
  queue = []
  begin
    1000.times do
      id = id_need_fetch.readline.chomp!
      queue << id unless id.empty?
    end
  rescue EOFError
    break
  end

  hydra = Typhoeus::Hydra.new(:max_concurrency => 10)
  reqs = []
  header = {"Accept-Encoding" => "gzip,deflate,sdch"}

  queue.each do |id|
    req = Typhoeus::Request.new("http://gdata.youtube.com/feeds/api/videos/#{id}/related?max-results=50&fields=entry(id)&alt=json",
                                :header => header)
    hydra.queue req
    reqs << req
  end

  log.info("fetch data")
  hydra.run

  ban = false
  reqs.each do |req|
    next unless req
    res = req.response
    if res.code == 200
      parse_related_id(res.body).each do |id|
        id_db.puts id.chomp
      end
      id_fetched.puts parse_id(res.effective_url)
    elsif res.code == 403 and res.body.index("too_many_recent_calls")
      #retry this urls next time
      queue << parse_id(res.effective_url).chomp
      ban = true
    else
      #other error, e.g. video not found
      id_fetched.puts parse_id(res.effective_url).chomp
    end
  end

  if ban
    log.info("sleep")
    sleep(300)
  end

end
